### PR TITLE
Initial development TcoCore 

### PR DESCRIPTION
**Pull request by PTKu**
**This is initial work and this PR aims to open discussion; all features here are still under development and undergoing heavy refactoring. Any input from the community is warmly welcomed.** Feel free to open a discussion here or in the **discussion tab above.**

# With this PR we move from Inxton organization to this repo and we will continue the development here

## Readme changes

https://github.com/Inxton/TcOpen/blob/dev/README.md

## Basic framework description

(updated) https://docs.tcopengroup.org/articles/TcOpenFramework/application.html

here is google document to comment on created by @GISED-Link 

**Please keep the discussion in the following document limited to the content, explanatory issues, etc** if you have suggestions regarding design, create and issue or discuss here https://github.com/TcOpenGroup/TcOpen/discussions/21

https://docs.google.com/document/d/1m0K2_1lwMMV4pcp_B2BamQrIaMImWBwfdyzMoOWHfa4/edit

## Documentation repository 

https://tcopengroup.github.io/TcOpen.Documentation/
||
https://docs.tcopengroup.org/

------------------------------
previous updates:
------------------------------

## TcoCore additions

although comitted by me most credit for the work on TcoCore goes to @TomKovac 

* TcoContext - main container for TcOpen application should provide application context
* TcoTask - coordination primitive that we called it in some place TIP (Task Interface Pattern)
* TcoState - rudimental state controller that provides components' state auto recovery
* TcoSequencer - more advanced controller for sequential logic
* Unit tests

**Documentation work is in progress that should help to get an idea about above mentioned stuff**

### Tc.Prober

we will keep it in this repo at this point to reduce dependency issues

### TcoIoBeckhoff

Should contain blocks for access and control of Beckhoff hardware. He we would like to discuss the architecture with the community.

### TcoPneumatics

Contains simple block for Pneumatic piston using TcoCore library + some unit tests

# Update
There is lot going on in this PR (repository) recently so here we would like to layout some basic concepts of the work that we have  done so far. 

We are bringing here many concepts that we developed over time and that were successfully used in many of our applications. However there is a danger that we bring a solution that is constrained to the scope and narrow paradigm that corresponds to the application domain specific to our applications. Therefore we would like to have a discussion where other ideas can improve the framework and make it suitable for wider variety of application domains.

We believe that in order to write components in OOP paradigm we need to create a set of basic classes that would allow for effective tasking, coordination, and state control of components as well as applications.   

## Project structure

Solution and projects are in VS2019. Projects are grouped (see bellow) each group has its own filtered solution 'slnf'. Filtered solution considerably speed up VS operations, this will come handy as TcOpen grows.

There are several TcProject, C#, WPF and testing project in this repository. The projects are grouped into these main chunks:

.\src

- **TcoCore** core construction blocks - Here is the main our focus at this point.
- **TcoIoBeckhoff** will contain classes for basic Beckhoff hardware
- **TcoPneumatics** component for pneumatic, we will be adding TcoDrive, TcoRobotics etc.
- **TcoApplicationExamples** examples, usage etc

C# and WPF project are there to help debug and test the libraries, they use inxton framework libraries.
C# project with 'Connector' suffix mediate PLC program for .net environment (WPF, Tests etc). 

## CI/CD

In ./pipeline folder you can find scripts and configuration for automated build.

## Documentation

We are working on 'documentation-as-code' using **docfx**. It will be in a separate repository here https://github.com/TcOpenGroup/TcOpen.Documentation

### TcContext : ITcoContext

Is an abstract class that should encapsulate a coherent unit (be it technology, station, more complex corelating parts of a device). It implements abstract method *Main* that is entry point of any logic of that particular unit. In default implementation it contains a counters that provide awareness for the containing objects about the state of execution, optional autonomous state restoration and possibly other features.

### TcoObject : ITcoObject

TcoObject is a class from which all other classes of the application should derive. TcoObject implements a series of diagnostics functions, support for messaging and reporting (HMI, alarms, debug etc.). It contains reference to its context as well as the reference of its parent object. It has the ability to query context and parent information.

### TcoTask : ITcoTask

TcoTask is the class that is capable of executing tasks in a asynchronous way (similar to Command pattern and Task Asynchronous Pattern). It has the ability to maintain done state and auto-restore/initialize on next invoke call (kind of auto-generated rising edge of the execute function).

### TcoComponent : ITcoComponent

TcoComponent base class for any component (drive, cylinder etc.). It is an abstract class that requires implementation of not only core logic, but manual/maintenance mode as well. 

### TcoState : ITcoState

TcoState is a elementary state controller that provides ability to auto-restore its components.

### TcoSequencer : ITcoSequencer

TcoSequencer is an advances state controller with higher observability, step-by-step operations, go-to step functions.

### CODE REVIEW

**Comment by jozefchmelar** [comment]
There's a lot going on in this PR.  but i understand. :)

There's A LOT OF tests...which is great.

I get the idea of Context, Component, Message, Task  and State...

eMessageCategory - There's a lot of clutter due to the documentation comment. Enums are self-explanatory.

TcoSelector is disabled, is it work in progress? Does it have to be there?


I would appreciate more practical examples, or a starter project with usage examples - ie. sequencer with pneumatics 

To sum it up : It has structure, it makes sense, it has plenty of test, interfaces are clearly defined ...well done

**Comment by PTKu** [comment]
@jozefchmelar messaging is unfinished... will need to work on it... that another concept that I would like to discuss later (https://github.com/Inxton/TcOpen/issues/12)
@TomKovac is working on on documentation and will provide some explanatory examples in short time

**Review by jozefchmelar — DISMISSED** [review]

**Comment by jozefchmelar** [comment]
There's again  a lot going on in this PR...Plenty of commits without any background. No issue, no task...I don't even know what am I supposed to review.

To run the code you need to do following

1 - Build TcoCoreConnector project
2 - Use the inxton extension or run the script .\_Vortex\build.ps1

All tests are passing locally 👍 

Sandboxes are working really well.

**Comment by Roald87** [comment]
>There's again a lot going on in this PR...Plenty of commits without any background. No issue, no task...I don't even know what am I supposed to review.

Agreed.

**Comment by PTKu** [comment]
> > There's again a lot going on in this PR...Plenty of commits without any background. No issue, no task...I don't even know what am I supposed to review.
> 
> Agreed.

I do agree too :). I do not want you guys to spend too much time reviewing here; we will need to squash it to something more concise and reviewable... as mentioned here https://github.com/TcOpenGroup/TcOpen/pull/13#issue-577804691 it is work in progress.. for people who like it somewhere between medium to raw :) really raw is here https://github.com/Inxton/TcOpen/tree/init-dev :)

**Comment by PTKu** [comment]
OK so...

There is lot going on in this PR (repository) recently so here we would like to layout some basic concepts of the work that we have  done so far. 

We are bringing here many concepts that we developed over time and that were successfully used in many of our applications. However there is a danger that we bring a solution that is constrained to the scope and narrow paradigm that corresponds to the application domain specific to our applications. Therefore we would like to have a discussion where other ideas can improve the framework and make it suitable for wider variety of application domains.

We believe that in order to write components in OOP paradigm we need to create a set of basic classes that would allow for effective tasking, coordination, and state control of components as well as applications.   

## Project structure

Solution and projects are in VS2019. Projects are grouped (see bellow) each group has its own filtered solution 'slnf'. Filtered solution considerably speed up VS operations, this will come handy as TcOpen grows.

There are several TcProject, C#, WPF and testing project in this repository. The projects are grouped into these main chunks:

.\src

- **TcoCore** core construction blocks - Here is the main our focus at this point.
- **TcoIoBeckhoff** will contain classes for basic Beckhoff hardware
- **TcoPneumatics** component for pneumatic, we will be adding TcoDrive, TcoRobotics etc.
- **TcoApplicationExamples** examples, usage etc

C# and WPF project are there to help debug and test the libraries, they use inxton framework libraries.
C# project with 'Connector' suffix mediate PLC program for .net environment (WPF, Tests etc). 

## CI/CD

In ./pipeline folder you can find scripts and configuration for automated build.

## Documentation

We are working on 'documentation-as-code' using **docfx**. It will be in a separate repository here https://github.com/TcOpenGroup/TcOpen.Documentation

## TcoCore library

### TcContext : ITcoContext

Is an abstract class that should encapsulate a coherent unit (be it technology, station, more complex corelating parts of a device). It implements abstract method *Main* that is entry point of any logic of that particular unit. In default implementation it contains a counters that provide awareness for the containing objects about the state of execution, optional autonomous state restoration and possibly other features.

### TcoObject : ITcoObject

TcoObject is a class from which all other classes of the application should derive. TcoObject implements a series of diagnostics functions, support for messaging and reporting (HMI, alarms, debug etc.). It contains reference to its context as well as the reference of its parent object. It has the ability to query context and parent information.

### TcoTask : ITcoTask

TcoTask is the class that is capable of executing tasks in a asynchronous way (similar to Command pattern and Task Asynchronous Pattern). It has the ability to maintain done state and auto-restore/initialize on next invoke call (kind of auto-generated rising edge of the execute function).

### TcoComponent : ITcoComponent

TcoComponent base class for any component (drive, cylinder etc.). It is an abstract class that requires implementation of not only core logic, but manual/maintenance mode as well. 

### TcoState : ITcoState

TcoState is a elementary state controller that provides ability to auto-restore its components.

### TcoSequencer : ITcoSequencer

TcoSequencer is an advances state controller with higher observability, step-by-step operations, go-to step functions.


just a reminder this PR is from fork https://github.com/Inxton/TcOpen

For the development we use the fork repository in inxton organization for two reasons
- to give us and the community freedom to undertake different path in the future
- we may need to keep the compatibility with existing code base that could potentially limit the community in further development

## most recent stuff is happening here https://github.com/Inxton/TcOpen/tree/init-dev

**Comment by GISED-Link** [comment]
TcoIoBeckhoff -> TcHalBeckhoff
HAL -> Hardware Abstraction Layer (commonly used in the embedded world)

I will take a serious look during the fallowing days

**Comment by PTKu** [comment]
> TcoIoBeckhoff -> TcHalBeckhoff
> HAL -> Hardware Abstraction Layer (commonly used in the embedded world)
> 
> I will take a serious look during the fallowing days

@GISED-Link 
There is a discussion opened around that topic here: https://github.com/TcOpenGroup/TcOpen/discussions/15

**Comment by Barteling** [comment]
awesome work guys, good job, it is to much a job for me right now for have a look at everything, but from a quick look it's very well coded.

A general thought what came to my mind, perhaps more a general discussion point than a PR review but anyway:

When the TcOpen initiative was started by Dustin I thought, alright nice we are gonna develop some general Function blocks, interfaces , a little bit like the Oscat libraries, general cylinder function blocks etc, easy to use, easy to understand. 

What we are heading to now starts to look like a full fledged framework. And I would need to take some serious looks at all the documentation and code to see what is going on. I think this might withhold less experienced programmers from starting to use the TcOpen libraries. and another point to consider , or something I would consider as company, is that once you start using frameworks, there is almost no way back. 

How do the others think about that?

**Comment by PTKu** [comment]
> A general thought what came to my mind, perhaps more a general discussion point than a PR review but anyway:
> 
> When the TcOpen initiative was started by Dustin I thought, alright nice we are gonna develop some general Function blocks, interfaces , a little bit like the Oscat libraries, general cylinder function blocks etc, easy to use, easy to understand.
> 
> What we are heading to now starts to look like a full fledged framework. And I would need to take some serious looks at all the documentation and code to see what is going on. I think this might withhold less experienced programmers from starting to use the TcOpen libraries. and another point to consider , or something I would consider as company, is that once you start using frameworks, there is almost no way back.
> 
> How do the others think about that?

Thanks, @Barteling for bringing these points into the discussion.... it is certainly something we should sort out at this point, so we better pause and go through it.

All your points are valid. 

I will argue for a framework that should be:

1. OOP oriented
1. Hide complexities, expose easy-to-use interfaces
1. Lead to good design 
1. Take some well-known burdens from the programmers (messaging-diagnostics, coordination, observability, etc.)
1. Be as much declarative as the language allows it to be (consumer code expresses ```what``` not ```how```).

As it is now components should not require the use of the whole framework except for:

1. There must be at least one implementation of  ```Context```, with ```Main``` method called at an appropriate place in the program
1. The components must take the ```Context``` as a parameter at construction (FB_init).

In this case, there would be some loss of functionality from the framework, but the components should be working.

In general, I believe that we can deliver better components if we use a framework.

As per a company selecting a framework... and getting stuck with it... yes this is what happens. A framework can be beneficial or harmful it depends on many factors. We should aim to create a framework that is worth using and the people and companies can benefit from it.

You're also right about less experienced programmers, being scared off. In my experience, if there are well-prepared templates for scaffolding the applications and examples the people will learn quickly. There is a certain group of automation engineers that will not venture into this kind of novelty for historical or ideological reasons. 

TwinCAT is a relatively young platform and there is a good chance there will be more people with software engineering background getting into the industry, we should aim in this direction. 

Few Monday morning thoughts on this topic 😄 

(@philippleidig @mark-lazarides @sagatowski  @Roald87)?

**Comment by sagatowski** [comment]
Hi! I have reviewed the documentation and here are some of my thoughts, attached as a word-document (copy-pasted text, and added comments directly in it).
[TcoCore review_js.docx](https://github.com/TcOpenGroup/TcOpen/files/6374563/TcoCore.review_js.docx)

**Comment by GISED-Link** [comment]
> Hi! I have reviewed the documentation and here are some of my thoughts, attached as a word-document (copy-pasted text, and added comments directly in it).
> [TcoCore review_js.docx](https://github.com/TcOpenGroup/TcOpen/files/6374563/TcoCore.review_js.docx)

Good idea!
I have import your document in google doc, so everyone could easily read and comment it without sharing a new version everytime. 
https://docs.google.com/document/d/1m0K2_1lwMMV4pcp_B2BamQrIaMImWBwfdyzMoOWHfa4/edit?usp=sharing

**Review by TomKovac — DISMISSED** [review]

**Comment by HAHermsen** [comment]
Well, the framework seems to provide complete solutions and a toolbox (let's call this a set of "design patterns") to build add or extend new functionalities in the standard provided solutions.

Plus I think most of Gerhard @Barteling's remarks are valid so I have thought of some idea's to gain even a wider acceptance;

If the set of design patterns pattern has a clear user guide (a first steps recipe type walkthrough) the acceptance level will raised considerably. 
I personally think that each provided FB should be usable in a "classic only" manner in CFC (without any OOP tricks involved!) for maximum acceptance by the widest audience.
The more experienced  users can also use the FB's in any extended OOP style according to their own experience level. This way the framework can also help people grasp certain OOP concepts steadily and gradually in their own tempo.

PS If you really wish to go agnostic consider building it without any Beckhoff specific dependencies. After all, ultimately Beckhoff's TwinCAT is just another flavor of CODESYS. We allready managed to "fork" TcUnit to a fully CODESYS compatible version named coUnit (formerly cfUnit) 
[https://forge.codesys.com/prj/cfunit/landing/Home/](https://forge.codesys.com/prj/cfunit/landing/Home/)

PS I have to read all documents thoroughly still (I just quickscanned them) and I am allready very busy on releasing coUnit v1.2 (Thank you @sagatowski ) and a release for coBolt, a SparkplugB implementation for CODESYS.

**Comment by PTKu** [comment]
reply to https://github.com/TcOpenGroup/TcOpen/pull/13#issuecomment-827544545

 > If the set of design patterns pattern has a clear user guide (a first steps recipe type walkthrough) the acceptance level will raised considerably.

Agreed! 

> I personally think that each provided FB should be usable in a "classic only" manner in CFC (without any OOP tricks involved!) for maximum acceptance by the widest audience.


> The more experienced users can also use the FB's in any extended OOP style according to their own experience level. This way the framework can also help people grasp certain OOP concepts steadily and gradually in their own tempo.

This would be great in theory. I am afraid it will lead to a convoluted design, though. We would be mixing procedural and object-oriented paradigms, it will inevitably create tensions in the design, which would reduce our possibilities considerably.  

As you said earlier we will need to bring guides, tutorials that will help the people to get on board. I would go that way.

> 
> PS If you really wish to go agnostic consider building it without any Beckhoff specific dependencies. After all, ultimately Beckhoff's TwinCAT is just another flavor of CODESYS. 

You are right we should bear this in mind.... at this point, there are some implementations that are tight to TwinCAT libraries (I guess Rtc implementation) There are more ways we can do this I would opt for abstractions with interfaces (then each platform can have its own implementation, where necessary).

**Comment by HAHermsen** [comment]
> This would be great in theory. I am afraid it will lead to a convoluted design, though. We would be mixing procedural and object-oriented paradigms, it will inevitably create tensions in the design, which would reduce our possibilities considerably.

I think we you misunderstandd me, which is my fault, as I did not explain too well what I really ment... Sorry for this.

What I am trying to say is that every FB should "usable" in a classic only manner and/or via OOP style approach.
This means that the FB's themselves indeed can (and should) be built in an OOP approach. Basicly mixed (classic/OOP) style programming is just adding VAR_INPUT / VAR_OUTPUT / VAR_IN_OUT parameters to FB's together with exposing various Interfaces.

For example, take a look at the commercial MQTT library by Janz-Tec, which follows this mixed approach.

**Comment by PTKu** [comment]
> > This would be great in theory. I am afraid it will lead to a convoluted design, though. We would be mixing procedural and object-oriented paradigms, it will inevitably create tensions in the design, which would reduce our possibilities considerably.
> 
> I think we you misunderstandd me, which is my fault, as I did not explain too well what I really ment... Sorry for this.
> 
> What I am trying to say is that every FB should "usable" in a classic only manner and/or via OOP style approach.
> This means that the FB's themselves indeed can (and should) be built in an OOP approach. Basicly mixed (classic/OOP) style programming is just adding VAR_INPUT / VAR_OUTPUT / VAR_IN_OUT parameters to FB's together with exposing various Interfaces.
> 
> For example, take a look at the commercial MQTT library by Janz-Tec, which follows this mixed approach.


@HAHermsen I'd like to have a look at the design of the MQTT library by Janz-Tec, just I do not have many options but to buy it just to look at the public interface.  
Could you prototype a simple component that would present that idea?

**Comment by PTKu** [comment]
Merged by other means #65
